nvflare/fuel/f3/streaming/stream_utils.py

```diff
@@ -34,8 +35,59 @@ def wrap_view(buffer: BytesAlike) -> memoryview:
     return view
 
 
-def gen_stream_id():
-    global lock, stream_count, start_time
+def gen_stream_id() -> int:
+    global lock, stream_count, sid_base
     with lock:
         stream_count += 1
-    return f"SID{(start_time + stream_count):16.0f}"
+    return sid_base + stream_count
+
+
+class FastBuffer:
+    """A buffer with fast appending"""
+
+    def __init__(self, buf: BytesAlike = None):
+        if not buf:
+            self.capacity = 1024
+        else:
+            self.capacity = len(buf)
+
+        self.buffer = bytearray(self.capacity)
+        if buf:
+            self.buffer[:] = buf
+            self.size = len(buf)
+        else:
+            self.size = 0
+
+    def to_bytes(self) -> BytesAlike:
+        """Return bytes-like object.
+        Once this method is called, append() may not work any longer, since the buffer may have been exported"""
+
+        if self.capacity == self.size:
+            result = self.buffer
+        else:
+            view = wrap_view(self.buffer)
+            result = view[0 : self.size]
+
+        return result
+
+    def append(self, buf: BytesAlike):
+        """Fast append by doubling the size of the buffer when it runs out"""
+
+        if not buf:
+            return self
+
+        length = len(buf)
+        remaining = self.capacity - self.size
+        if length > remaining:
+            # Expanding the array as least twice the current capacity
+            new_cap = max(length + self.size, 2 * self.capacity)
+            self.buffer = self.buffer.ljust(new_cap, b"\x00")
+            self.capacity = new_cap
+
+        self.buffer[self.size :] = buf
+        self.size += length
+
+        return self
```

Comment:
I thought python array is already a dynamic array with some growth factor approx to 1.125 (https://en.wikipedia.org/wiki/Dynamic_array#Growth_factor).

Do we just make this 2 so it grows faster?

So the growth factor is the key to the ACK timeout issue?

Reply:
Correct. The number of allocations is the key.

The growth factor is only for list, not bytearray.